docs/docs/specs/v0/composer.md

```diff
@@ -0,0 +1,190 @@
+
+```ts
+type MaybePromise<T> = T | Promise<T>;
+
+type Plugin = {
+  meta: {
+    // serializable, could be stored in a space.
+    id: string;
+    name: string;
+    description: string;
+  };
+  provides: {
+    tree: {
+      getTreeNodes(parent: TreeNode): Observable<TreeNode[]>;
+      getTreeNodeActions(parent: TreeNode): Observable<Action[]>;
+    };
+    content: {
+      getComponent(selection: TreeNode[]): MaybePromise<React.FC>;
+    };
+  };
+};
+
+type Effect = (state: ComposerState) => MaybePromise<ComposerState>;
+
+type Action = {
+  id: string;
+  label: string;
+  icon?: React.FC;
+  invoke(state: ComposerState): MaybePromise<Effect | Effect[]>;
+};
+
+type TreeNode<T = any> = {
+  id: string;
+  data?: T;
+  label: string;
+  icon?: React.FC;
+  actions?: Action[];
+  loading?: boolean;
+  disabled?: boolean;
+  children?: TreeNode[];
+  labelEditable?: boolean;
+  onLabelChanged?(value: string): any;
+};
+
+type ComposerState = {
+  location: string;
+  surfaces: {
+    tree: {
+      selection: TreeNode[];
+      nodes: TreeNode[];
+    };
+    sidebar: {
+      isOpen: boolean; // is the sidebar currently open
+      isPinned: boolean; // whether the sidebar will autohide
+    };
+  };
+  plugins: Plugin[];
+  client: Client;
+  // etc
+};
+```
+
+In order to populate the tree, plugins are first asked to present their lists of children without a `parent` node (or a stand-in root node value). This generates the first level items in the Tree. Then, for each node ad-nauseum, plugins are asked to return more children until the tree reaches a steady state. This allows plugins to add nodes to each other's nodes.
+
+Things to think about:
+
+- how to do paging of large result sets
+- how to detect circular / infinite trees and deal with them
+- how to expand `getTreeNodes` lazily / in a timely manner without losing too much fidelity in the Tree
+
+Some of the first plugins:
+
+0. the spaces plugin - which provides a list of root nodes representing accessible ECHO spaces
+1. the markdown plugin - which provides a plain text editor for the content area and fills the Tree with document and folder nodes
+2. the filesystem plugin - which provides import / export to folders on disk
+3. the github plugin - which provides nodes representing github issues and assets
+4. the stacks plugin - which provides a stack editor for the content area and import / export actions to github
+
+If a stack with custom tiles (frames) is required, that is just an extension of the stacks plugin, where the `Stack` returned from `getComponent` is endowed with more kinds of frames statically.
+
+#### How Kai relates to this model
+
+In v1 (static chrome) Kai can take advantage of the composer chrome to get ahead on compatibility with mobile screens, take advantage of the magic search box, and reduce boilerplate. Kai can re-implement itself as a specific expression of the `<Composer />` element with custom values for the sidebar content and list of plugins.
```

Comment:
This is helpful re short-term convergence of Kai/Composer at the low-level component layer (e.g., common shell).
However, as discussed elsewhere, there's a large overlap (but not congruence) between the concepts of Frames explored and implemented in kai and Composer-specific plugins. Given the ambition of what's being laid out here, we should further tease out core/universal "Braneframe" framework concepts from v1/v2 Composer/Kai functionality (i.e., rather than Composer "the app" driving core framework decisions: e.g., trees, search, etc.)

That is to say, Composer can be the "driver" of focus of the framework, but should stand-alone from the framework conceptual organization. Otherwise we may overspecialize the framework.

Reply:
good discussion point.
in the meantime, I renamed `ComposerState` with `AppState`, illustrated Frames in a glossary, as well as outlined key differences between Kai frames and Composer extensibility at the bottom